### PR TITLE
[WIP] nspawn: Add --syscall-filter option

### PR DESCRIPTION
**Pull request by robertgzr**
Adds a --syscall-filter option to modify nspawns default seccomp filter.
After the comments in #5944 this now works similar to the `SystemCallFilter=`
option in unit files, but only accepts system calls.

This option takes a comma-separated list of system calls to add to the default filter.
If the first character of the list is "~", the listed system calls will be whitelisted
instead. I modify the blacklist before handling the `--[drop-]capability` option,
so anything there will overwrite this option.

Fixes #5163

### CODE REVIEW

File: src/nspawn/nspawn-seccomp.c [hunk 1]
```diff
@@ -29,111 +29,154 @@
...
                         continue;
 
-                r = seccomp_rule_add_exact(ctx, SCMP_ACT_ERRNO(EPERM), blacklist[i].syscall_num, 0);
-                if (r < 0) {
-                        /* If the system call is not known on this architecture, then that's fine, let's ignore it */
-                        _cleanup_free_ char *n = NULL;
+                r = seccomp_filter_set_add(blacklist, !whitelist, &(cap_filter_sets[i].set));
+                if (r < 0)
+                        log_debug_errno(r, "Failed to add capability to blacklist, ignoring: %m");
+                /* NULSTR_FOREACH(t, cap_filter_sets[i].set.value) { */
```
Comment (robertgzr) [inline src/nspawn/nspawn-seccomp.c]: I wasn't sure if the custom filter should also whitelist syscalls from the capability filter sets. I'm not sure what the priorities are there.

**Comment by dongsupark** [comment]
Nit: you should not write `Signed-off-by` in your commit message in systemd.

File: src/shared/seccomp-util.h [hunk 2]
```diff
@@ -69,6 +69,7 @@ const SyscallFilterSet *syscall_filter_set_find(const char *name);
 
 int seccomp_filter_set_add(Set *s, bool b, const SyscallFilterSet *set);
 
+int seccomp_add_syscall_filter_set(scmp_filter_ctx seccomp, uint32_t default_action, const SyscallFilterSet *set, uint32_t action);
```
Comment (iaguis) [inline src/shared/seccomp-util.h]: Why are we exporting this? Doesn't seem to be used outside of `src/shared/seccomp-util.c`

**Comment by robertgzr** [comment]
@dongsupark 👍

File: src/nspawn/nspawn-seccomp.c [hunk 3]
```diff
@@ -29,111 +29,154 @@
...
-                uint64_t cap_list_retain) {
+static const SyscallFilterSet nspawn_default_blacklist = {
+        .name = "@nspawn_default_blacklist",
+        .help = "Syscall filter for systemd-nspawn",
+        .value =
+        "_sysctl\0"
+        "add_key\0"
+        "afs_syscall\0"
+        "bdflush\0"
+        "bpf\0"
```
Comment (iaguis) [inline src/nspawn/nspawn-seccomp.c]: This should be wrapped by 

```
#ifdef __NR_bpf
#endif
```

right?

File: src/nspawn/nspawn-seccomp.c [hunk 4]
```diff
@@ -29,111 +29,154 @@
...
+        "break\0"
+        "create_module\0"
+        "ftime\0"
+        "get_kernel_syms\0"
+        "getpmsg\0"
+        "gtty\0"
+        "kexec_file_load\0"
```
Comment (iaguis) [inline src/nspawn/nspawn-seccomp.c]: same ifdef thing

File: src/nspawn/nspawn-seccomp.c [hunk 5]
```diff
@@ -29,111 +29,154 @@
...
+}
+
+static int seccomp_setup_filter(
+                scmp_filter_ctx ctx,
+                Set *blacklist,
+                uint64_t cap_list_retain,
+                Set *syscall_filter,
+                bool whitelist
+                ) {
+
```
Comment (iaguis) [inline src/nspawn/nspawn-seccomp.c]: superfluous empty line

File: src/nspawn/nspawn-seccomp.c [hunk 6]
```diff
@@ -29,111 +29,154 @@
...
 
-                r = seccomp_rule_add_exact(ctx, SCMP_ACT_ERRNO(EPERM), blacklist[i].syscall_num, 0);
-                if (r < 0) {
-                        /* If the system call is not known on this architecture, then that's fine, let's ignore it */
-                        _cleanup_free_ char *n = NULL;
+                r = seccomp_filter_set_add(blacklist, !whitelist, &(cap_filter_sets[i].set));
+                if (r < 0)
+                        log_debug_errno(r, "Failed to add capability to blacklist, ignoring: %m");
+                /* NULSTR_FOREACH(t, cap_filter_sets[i].set.value) { */
+                /*         if (syscall_filter && whitelist && set_contains(syscall_filter, t)) { */
```
Comment (iaguis) [inline src/nspawn/nspawn-seccomp.c]: Does this work for removing `@` groups?

File: src/nspawn/nspawn.c [hunk 7]
```diff
@@ -1051,6 +1058,54 @@ static int parse_argv(int argc, char *argv[]) {
...
+                                optarg++;
+                        }
+
+                        if (!arg_syscall_filter) {
+                                arg_syscall_filter = set_new(NULL);
+                                if (!arg_syscall_filter)
+                                        return log_oom();
+
+                                if (invert)
+                                        /* Apply default filter and the ones listed */
```
Comment (iaguis) [inline src/nspawn/nspawn.c]: isn't this the other way around? if invert we're whitelisting so we're excluding the syscalls listed from the filter

File: src/nspawn/nspawn.c [hunk 8]
```diff
@@ -1051,6 +1058,54 @@ static int parse_argv(int argc, char *argv[]) {
...
+                                if (r == 0)
+                                        break;
+                                if (r == -ENOMEM)
+                                        return log_oom();
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to parse syscall whitelist: %s", optarg);
+
+                                r = set_put(arg_syscall_filter, t);
+                                if (r == 0)
+                                        return 0;
```
Comment (iaguis) [inline src/nspawn/nspawn.c]: why are we returning here? IIUC, when `set_put()` returns 0 it means the key already exists so we can just ignore and continue the loop

File: src/nspawn/nspawn-seccomp.c (hunk 1 above)
Comment (iaguis) [reply src/nspawn/nspawn-seccomp.c]: One angle would be that we should allow to whitelist those syscalls so apps can get a proper EACCESS or whatever instead of the seccomp blocking behavior, whatever that is.

Other angle would be that if the user didn't want to keep certain capability, restricting the syscalls adds a bit more security in the case there's a bug in the capabilities implementation.

I think that if the user doesn't enable a capability but enables the syscalls it's their fault and they should be allowed to shoot themselves on the foot :)

TL;DR whitelisting is fine IMO

**Review by iaguis — CHANGES_REQUESTED** [review]
Some comments.

File: src/shared/seccomp-util.h (hunk 2 above)
Comment (robertgzr) [reply src/shared/seccomp-util.h]: Leftover from the first PR. Removed.

File: src/nspawn/nspawn.c (hunk 7 above)
Comment (robertgzr) [reply src/nspawn/nspawn.c]: The comment was wrong 😛 

I was going by this https://www.freedesktop.org/software/systemd/man/systemd.exec.html#SystemCallFilter=
And there it acts as a whitelist by default and as a blacklist when you invert with `~`

File: src/nspawn/nspawn-seccomp.c (hunk 3 above)
Comment (robertgzr) [reply src/nspawn/nspawn-seccomp.c]: 👍

File: src/nspawn/nspawn-seccomp.c (hunk 4 above)
Comment (robertgzr) [reply src/nspawn/nspawn-seccomp.c]: 👍

File: src/nspawn/nspawn-seccomp.c [hunk 9]
```diff
@@ -29,111 +29,155 @@
...
+                bool iswhitelist
+                ) {
+        int r;
+
+        /* parse the default blacklist */
+        r = seccomp_filter_set_add(blacklist, true, &nspawn_default_blacklist);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to add default filter to blacklist: %m");
+
+        /* Filter out syscalls by capabilities but retain the ones on the list */
```
Comment (robertgzr) [inline src/nspawn/nspawn-seccomp.c:137]: @iaguis I add the "capability syscalls" to the blacklist first...

File: src/nspawn/nspawn-seccomp.c [hunk 10]
```diff
@@ -29,111 +29,155 @@
...
-                        log_debug_errno(r, "Failed to add rule for system call %s, ignoring: %m", strna(n));
-                } else
-                        c++;
+        /* apply the custom filter */
+        if (syscall_filter) {
+                Iterator i;
+                char *t;
+
+                SET_FOREACH(t, syscall_filter, i) {
+                        if (t[0] == '@') {
```
Comment (robertgzr) [inline src/nspawn/nspawn-seccomp.c:148]: ... and it now supports filter sets

File: src/nspawn/nspawn-seccomp.c [hunk 11]
```diff
@@ -29,111 +29,155 @@
...
-
-        static const struct {
-                uint64_t capability;
-                int syscall_num;
-        } blacklist[] = {
-                { 0,              SCMP_SYS(_sysctl)             }, /* obsolete syscall */
-                { 0,              SCMP_SYS(add_key)             }, /* keyring is not namespaced */
-                { 0,              SCMP_SYS(afs_syscall)         }, /* obsolete syscall */
-                { 0,              SCMP_SYS(bdflush)             },
+static const SyscallFilterSet nspawn_default_blacklist = {
```
Comment (robertgzr) [inline src/nspawn/nspawn-seccomp.c:41]: I would move this into here 
https://github.com/kinvolk/systemd/blob/master/src/shared/seccomp-util.c#L270
and document it here 
https://www.freedesktop.org/software/systemd/man/systemd.exec.html#SystemCallFilter=

This would make it much easier to find the reason for the bug we had originally.
Going to ask that in the PR.
